watchdog.go

```diff
@@ -148,72 +184,113 @@ func (w *Watchdog) Start() {
 
 	reaploop:
 		for {
+			w.mut.Lock()
+			if w.retryCount > w.MaxRetries {
+				w.logEvent(true, "unable to start after %v retries, giving up", w.retryCount)
+				w.err = fmt.Errorf("unable to start process after %v retries, giving up", w.retryCount)
+				return
+			}
+			w.mut.Unlock()
+
 			if w.needRestart {
-				if w.proc != nil {
-					w.proc.Release()
+				if w.cmd != nil && w.cmd.Process != nil {
+					w.cmd.Process.Release()
 				}
-				Q(" debug: about to start '%s'", w.PathToChildExecutable)
-				w.proc, err = os.StartProcess(w.PathToChildExecutable, w.Args, &w.Attr)
+				w.logEvent(true, "about to start '%s'", w.PathToChildExecutable)
+				//w.cmd.SysProcAttr = &w.Attr;
+
+				w.mut.Lock()
+				if w.retryCount > 0 {
+					w.logEvent(true, "sleeping for %v before attempting restart; retryCount = %v (max = %v)", w.RetryInterval, w.retryCount, w.MaxRetries)
+					time.Sleep(w.RetryInterval)
+				}
+				w.mut.Unlock()
+
+				w.cmd = exec.Command(w.PathToChildExecutable, w.Args...)
+				w.cmd.Dir = w.Cwd
+
+				w.cmd.Stdout = &w.sout
+				w.cmd.Stderr = &w.serr
+
+				w.Restarting <- true
```

Comment:
`w.Restarting <- true` Could block if the client isn't tending to events promptly.
Safety would suggest making that impossible 
~~~
select {
 w.Restarting <- true
 default:
}
~~~

Reply:
Yes that is true, however, going with your solution, the outside might not be alerted that a restart is taking place.
We have had to fork the repo, since we wanted watchdog to inform when a restart is happening. (since we need to do some re-subscription of some events attached to the process). Perhaps you had another way of handling that ?

Reply:
> the outside might not be alerted that a restart is taking place.

Huh?  Why not?  The channel has 1 buffer space in it, so it will always get filled if empty.

Reply:
you're right on that one.

Was there originally a way to notify the outside of a restart ?

Reply:
I think your fork and restart mechanism are fine.

I don't recall any other means; the restart seems fine with the minor repair suggested.

 NB: Although I wrote bark, its under the betable acct. Since I don't work there anymore so I don't have access to handle pull requests. Sorry about that. Nothing I can do about it really. I have a fork on my personal account if you want to suggest pulls to that.

Reply:
the fork on this repo was a mistake, that's why I closed it straightaway. I would be interested in your personal fork, thanks!